ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V100__Append_only_schema.sql

```diff
@@ -319,6 +315,240 @@ CREATE INDEX participant_events_non_consuming_exercise_template_id_idx ON partic
+    participant_events.transaction_id,
+    participant_events.ledger_effective_time,
+    participant_events.command_id,
+    participant_events.workflow_id,
+    participant_events.application_id,
+    participant_events.submitters,
+    participant_events.node_index,
+    participant_events.event_id,
+    participant_events.contract_id,
+    participant_events.template_id,
+    participant_events.flat_event_witnesses,
+    participant_events.tree_event_witnesses,
+    participant_events.create_key_value,
+    participant_events.create_key_value_compression,
+    participant_events.exercise_choice,
+    participant_events.exercise_argument,
+    participant_events.exercise_argument_compression,
+    participant_events.exercise_result,
+    participant_events.exercise_result_compression,
+    participant_events.exercise_actors,
+    participant_events.exercise_child_event_ids
+FROM participant_events
+WHERE participant_events.exercise_consuming = FALSE -- non-consuming exercise events
+ORDER BY participant_events.event_sequential_id;
+
+-- Temporary sequence to generate sequential IDs that appear after
+-- all other already existing sequential IDs.
+CREATE SEQUENCE temp_divulgence_sequential_id START 1;
+
+-- Divulgence events did not exist before, we need to assign a new sequential ID for them.
+-- They will all be inserted after all other events, i.e., at a point later than the transaction
+-- that actually lead to the divulgence. This is OK, as we only use the for lookups
+-- AND we can recognize these divulgence events from not having an associated ???.
+-- In addition, we want to avoid rewriting the event_sequential_id of other events
+-- for data continuity reasons.
+WITH divulged_contracts AS (
+    SELECT nextval('temp_divulgence_sequential_id') +
+           (SELECT coalesce(max(event_sequential_id), 0) FROM participant_events) as event_sequential_id,
+           contract_id,
+           array_agg(contract_witness) as divulgees
+    FROM participant_contract_witnesses
+             INNER JOIN participant_contracts USING (contract_id)
+             LEFT JOIN participant_events USING (contract_id)
+    WHERE (NOT create_stakeholders @> array [contract_witness])
+      AND exercise_consuming IS NULL -- create events only
+      AND (tree_event_witnesses IS NULL OR NOT tree_event_witnesses @> array [contract_witness])
+    GROUP BY contract_id
+) INSERT INTO participant_events_divulgence (
+    event_sequential_id,
+    event_offset,
+    command_id,
+    workflow_id,
+    application_id,
+    submitters,
+    contract_id,
+    template_id,
+    tree_event_witnesses,
+    create_argument,
+    create_argument_compression
+)
+SELECT
+    event_sequential_id,
+    -- The following 5 fields are metadata of the transaction that lead to the divulgence.
+    -- We can't reconstruct this information from the old schema.
+    NULL, -- TODO append-only: this is the only reason why the event_offset is nullable. Use some non-null offset instead?
```

Comment:
```suggestion
    NULL,
```
Once we restore from a snapshot of the data on the committer we're also not going to have an offset available. I'd suggest we just leave it nullable, and deal with that fact in our code. Conceptually, we should consider the event_sequential_id to be the main ordering domain, and the offsets to be transaction metadata, which we might not have available for contracts created before `pruning_end`.

Reply:
Thanks for the hint! I have removed the comment. The event offset for create events is still non-nullable, I suggest we make it nullable when we introduce snapshot restore, or in a subsequent PR (in order to keep this PR small).

Reply:
Agreed.